### PR TITLE
feat: adding identity pool credentials

### PR DESCRIPTION
**Pull request by BigTailWolf**
Add a new type of external account `Google::Auth::ExternalAccount::IdentityPoolCredentials` which reads the config of workforce/workload identity pool.
This type of credentials retrieve OIDC token (subject token) to exchange with STS (access_token).

### CODE REVIEW

**Review by bajajneha27 — COMMENTED** [review]
LGTM overall, just a few nit comments

File: spec/googleauth/external_account/aws_request_signer_spec.rb [hunk 1]
```diff
@@ -465,7 +465,7 @@
     it "fulfils Amazon's test \##{index}" do
       allow(Time).to receive(:now).and_return(Time.strptime(fixture[:time], '%Y-%m-%dT%H:%M:%SZ'))
       request_signer = AwsRequestSigner.new(fixture[:region])
-      actual_signed_request = request_signer.generate_signed_request(fixture[:credentials], fixture[:original_request])
```
[the remark this answers is not in the document]
Comment (BigTailWolf) [reply spec/googleauth/external_account/aws_request_signer_spec.rb:468]: > Why are removing this?

I'm not removing that. Not sure what happened to github. Let me recheck

[the remark this answers is not in the document]
Comment (BigTailWolf) [reply spec/googleauth/external_account/aws_request_signer_spec.rb:468]: actual_signed_request = request_signer.generate_signed_request(fixture[:credentials], fixture[:original_request])

File: lib/googleauth/external_account/identity_pool_credentials.rb [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+          @credential_source = options[:credential_source] || {}
+          @credential_source_file = @credential_source[:file]
+          @credential_source_url = @credential_source[:url]
+          @credential_source_headers = @credential_source[:headers] || {}
+          @credential_source_format = @credential_source[:format] || {}
+          @credential_source_format_type = @credential_source_format[:type] || "text"
+          validate_credential_source
+        end
+
+        def retrieve_subject_token!
```
Comment (dazuma) [inline lib/googleauth/external_account/identity_pool_credentials.rb:44]: Missing documentation. All public methods need to be documented. (If this method is not meant to be part of the public interface, annotate it with `# @private`.)

File: lib/googleauth/external_account/identity_pool_credentials.rb [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      # This module handles the retrieval of credentials from Google Cloud by utilizing the any 3PI
+      # provider then exchanging the credentials for a short-lived Google Cloud access token.
+      class IdentityPoolCredentials
+        include Google::Auth::ExternalAccount::BaseCredentials
+        extend CredentialsLoader
+
+        # Will always be nil, but method still gets used.
+        attr_reader :client_id
+
+        def initialize options = {}
```
Comment (dazuma) [inline lib/googleauth/external_account/identity_pool_credentials.rb:31]: Missing documentation.

File: lib/googleauth/external_account/identity_pool_credentials.rb [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+          content, resource_name = token_data
+          if @credential_source_format_type == "text"
+            token = content
+          else
+            begin
+              response_data = MultiJson.load content, symbolize_keys: true
+              token = response_data[@credential_source_field_name.to_sym]
+            rescue StandardError
+              raise "Unable to parse subject_token from JSON resource #{resource_name}
+              using key #{@credential_source_field_name}"
```
Comment (dazuma) [inline lib/googleauth/external_account/identity_pool_credentials.rb:54]: Don't split the string across lines; currently this string includes the newline and indent whitespace, which is probably not what you want. If you need to provide a string literal that's too long for the line length limit, use the following syntax:

```ruby
              raise "Unable to parse subject_token from JSON resource #{resource_name} " \
                    "using key #{@credential_source_field_name}"
```

**Review by dazuma — CHANGES_REQUESTED** [review]

File: lib/googleauth/external_account/identity_pool_credentials.rb (hunk 2 above)
Comment (BigTailWolf) [reply lib/googleauth/external_account/identity_pool_credentials.rb:44]: It's implementation of `BaseCredentials` module's private method. Moved to private properly.

File: lib/googleauth/external_account/identity_pool_credentials.rb [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        def retrieve_subject_token!
+          content, resource_name = token_data
+          if @credential_source_format_type == "text"
+            token = content
+          else
+            begin
+              response_data = MultiJson.load content, symbolize_keys: true
+              token = response_data[@credential_source_field_name.to_sym]
+            rescue StandardError
+              raise "Unable to parse subject_token from JSON resource #{resource_name}" \
```
Comment (dazuma) [inline lib/googleauth/external_account/identity_pool_credentials.rb:62]: Space at the end of this string. The strings on this line and the next are concatenated directly, and you probably want a space between them, between the words in the sentence.

**Review by dazuma — APPROVED** [review]
LGTM

File: lib/googleauth/external_account/external_account_utils.rb [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+require "googleauth/oauth2/sts_client"
+
+module Google
+  # Module Auth provides classes that provide Google-specific authorization
+  # used to access Google APIs.
+  module Auth
+    module ExternalAccount
+      # Authenticates requests using External Account credentials, such
+      # as those provided by the AWS provider or OIDC provider like Azure, etc.
+      module ExternalAccountUtiles
```
Comment (bajajneha27) [inline lib/googleauth/external_account/external_account_utils.rb:26]:
```suggestion
      module ExternalAccountUtils
```

File: lib/googleauth/external_account/identity_pool_credentials.rb [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+module Google
+  # Module Auth provides classes that provide Google-specific authorization used to access Google APIs.
+  module Auth
+    module ExternalAccount
+      # This module handles the retrieval of credentials from Google Cloud by utilizing the any 3PI
+      # provider then exchanging the credentials for a short-lived Google Cloud access token.
+      class IdentityPoolCredentials
+        include Google::Auth::ExternalAccount::BaseCredentials
+        include Google::Auth::ExternalAccount::ExternalAccountUtiles
```
Comment (bajajneha27) [inline lib/googleauth/external_account/identity_pool_credentials.rb:27]:
```suggestion
        include Google::Auth::ExternalAccount::ExternalAccountUtils
```

**Pull request merged**